pip/pip-341.md

```diff
@@ -0,0 +1,305 @@
+        }
+    }
+
+    @Override
+    public void recordMessageReceived(Message<?> message) {
+        statsRecorder.updateNumMsgsReceived(message);
+    }
+
+    // Other methods...
+}
+```
+
+This allows us to completely replace all existing calls to the stat recorders with our metrics trackers, while still maintaining backwards compatibility with the previous behavior.
+
+### Backwards compatibility with statsIntervalSeconds configuration
+We can mark the `statsIntervalSeconds` client configuration as deprecated since implementation-specific configurations should be provided via the params map when constructing the metrics tracker in `ClientBuilder` (`statsIntervalSeconds` is specific to `LoggingMetricsTrackerFactory`).
+
+We will set the default metrics tracker factory as the `LoggingMetricsTrackerFactory` to maintain backwards compatibility, however we still need to do some checks:
+1. If a non-default non-null metrics tracker factory (ex. OTel one) is configured, then use it
+2. Else if `statsIntervalSeconds=0`, then the user has explicitly disabled client metrics so we must use the `NoOpMetricsTrackerFactory`
+3. Else, use `LoggingMetricsTrackerFactory`
+
+Inside `LoggingMetricsTrackerFactory`, we must also derive `statsIntervalSeconds`:
+1. First try to use `statsIntervalSeconds` if it is configured via the new params map in `ClientBuilder metricsTrackerFactory(String metricsTrackerFactoryClassName, Map<String, String> params)`
+2. Otherwise, try to use the deprecated `statsIntervalSeconds` value if it is greater than 0
+3. Finally, default to 60 seconds if `statsIntervalSeconds` is not configured anywhere
+
+### Supporting existing Consumer/Producer getStats API
+There are two major challenges with supporting the existing producer/consumer `getStats()` API:
+1. The returned object depends on whether the topic is partitioned or non-partitioned (`PartitionedTopicProducerStatsRecorderImpl`/`MultiTopicConsumerStatsRecorderImpl` vs `ProducerStatsRecorderImpl`/`ConsumerStatsRecorderImpl`). This is problematic for our metrics tracker interface as it means the user must know when to provide the partitioned vs non-partitioned stats recorder by checking the producer/consumer class type.
+2. The existing stat recorders do not provide setter/builder/supplier methods to construct a stats recorder with specific values (they only provide the increment methods). This is problematic as the user will have to maintain an updated stats recorder instance in the metrics tracker, which can result in duplicate event recording (one call to stats recorder, and another call to the user's implementation for each event).
+
+We can alleviate these issues by making two changes:
+1. Consolidate to a new stats supplier class to handle both partitioned & non-partitioned cases
+2. Add setter/builder/supplier methods to the new stats supplier class to allow the metrics tracker to supply stat values
+
+#### Consolidated ProducerStatsSupplier/ConsumerStatsSupplier
```

Comment:
What's the context here. Maybe you should explain in background knowledge what are those suppliers.